src/txmempool.cpp

```diff
@@ -395,9 +397,19 @@ void CTxMemPoolEntry::UpdateAncestorState(int32_t modifySize, CAmount modifyFee,
     assert(int(nSigOpCostWithAncestors) >= 0);
 }
 
-CTxMemPool::CTxMemPool(const Options& opts)
-    : m_opts{opts}
+static CTxMemPool::Options SetOptionLimits(CTxMemPool::Options&& opts)
```

Comment:
In commit "[[refactor]] Check CTxMemPool options in constructor" (27af3910e5fa9915d4913b82443db216d1b1d61b)

I think it would be clearer to avoid using && and std::move, since using these doesn't actually make things more efficient. In 27af3910e5fa9915d4913b82443db216d1b1d61b, one copy and one move of the option struct will happen. The copy happens when calling the CTxMemPool constructor (since the parameter type is a value parameter), the move happens when returning from SetOptionLimits.

But you could get the same result just using plain references and values:

```diff
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -397,19 +397,20 @@ void CTxMemPoolEntry::UpdateAncestorState(int32_t modifySize, CAmount modifyFee,
     assert(int(nSigOpCostWithAncestors) >= 0);
 }
 
-static CTxMemPool::Options SetOptionLimits(CTxMemPool::Options&& opts)
+//! Clamp option values and return error if options are not valid.
+static CTxMemPool::Options NormalizeOptions(CTxMemPool::Options opts, bilingual_str& error)
 {
     opts.check_ratio = std::clamp<int>(opts.check_ratio, 0, 1'000'000);
+    int64_t descendant_limit_bytes = opts.limits.descendant_size_vbytes * 40;
+    if (opts.max_size_bytes < 0 || opts.max_size_bytes < descendant_limit_bytes) {
+        error = strprintf(_("-maxmempool must be at least %d MB"), std::ceil(descendant_limit_bytes / 1'000'000.0));
+    }
     return opts;
 }
 
-CTxMemPool::CTxMemPool(Options opts, bilingual_str& error)
-    : m_opts{SetOptionLimits(std::move(opts))}
+CTxMemPool::CTxMemPool(const Options& opts, bilingual_str& error)
+    : m_opts{NormalizeOptions(opts, error)}
 {
-    int64_t descendant_limit_bytes = m_opts.limits.descendant_size_vbytes * 40;
-    if (m_opts.max_size_bytes < 0 || m_opts.max_size_bytes < descendant_limit_bytes) {
-        error = strprintf(_("-maxmempool must be at least %d MB"), std::ceil(descendant_limit_bytes / 1'000'000.0));
-    }
 }
 
 bool CTxMemPool::isSpent(const COutPoint& outpoint) const
```

This avoids need for std::move and &&, keeps the `CTxMemPool` constructor parameter unchanged, still does one copy and one move when calling and returning from NormalizeOptions. This also keeps error checking and clamping in the same function, which could be helpful if more options are added, because if clamping happens in one function and error checking happens in another function it could make the checking less flexible or require duplicate code.

Reply:
I was just re-using the pattern used for the `ChainstateManager` construction, but I like this more. Taking.

Reply:
>  Flatten avoid an unnecessary move:

What do you mean with this?

Reply:
> > Flatten avoid an unnecessary move:

> What do you mean with this?

I was suggesting returning `Options&&` instead of `Options` to be consistent with the validation.cpp flatten code, and to avoid two moves potentially happening on return: One move on the `return opts;` statement which the compiler can't elide because `opts` is not a local variable, and another move after the function returns to initialize `m_opts` member.

In practice, however, it looks like only one move will happen, because even though the compiler can't apply NRVO when returning `opts` it _can_ use RVO to initialize `m_opts` in-place. So in practice returning either `Options` or `Options&&` should be equivalent, and I could see reasons for choosing either approach.

I think I still might opt to copy the validation.cpp flatten name and types so the pattern is the same both places, while also adding a `string& error` output parameter so error checking can happen alongside flattening. But any approach you want to take seems fine to me. And thanks for asking what was wrong with returning an `Options` value, because I was incorrectly assuming it was less efficient than it really is.